include/jemalloc/internal/tcache_inlines.h

```diff
@@ -32,13 +32,65 @@ tcache_enabled_set(tsd_t *tsd, bool enabled) {
 	tsd_slow_update(tsd);
 }
 
+static inline unsigned
+thread_nhbins_get(tsd_t *tsd) {
+	return tsd_thread_nhbins_get(tsd);
+}
+
+static inline void
+thread_nhbins_set(tsd_t * tsd, unsigned thread_nhbins) {
+	if (thread_nhbins > TCACHE_NBINS_MAX) {
+		thread_nhbins = TCACHE_NBINS_MAX;
+	}
+	tsd_thread_nhbins_set(tsd, thread_nhbins);
+	size_t tcache_max = tsd_thread_tcache_max_get(tsd);
+	assert(thread_nhbins == sz_size2index(tcache_max) + 1);
+}
+
+static inline size_t
+thread_tcache_max_get(tsd_t *tsd) {
+	return tsd_thread_tcache_max_get(tsd);
+}
+
+static inline void
+thread_tcache_max_set(tsd_t *tsd, size_t tcache_max) {
+	/* tcache_max should not be larger than TCACHE_MAXCLASS_LIMIT. */
+	if (tcache_max > TCACHE_MAXCLASS_LIMIT) {
+		tcache_max = (size_t)TCACHE_MAXCLASS_LIMIT;
+	}
+	/* Ceil tcache_max to a valid size class. */
+	tcache_max = sz_s2u(tcache_max);
+
+	size_t old_tcache_max = tsd_thread_tcache_max_get(tsd);
+	if (old_tcache_max != tcache_max && tcache_available(tsd)) {
+		/* Reboot the tcache for a clean set of tcache_max. */
+		tcache_cleanup(tsd);
+		tsd_thread_tcache_max_set(tsd, tcache_max);
+		thread_nhbins_set(tsd, sz_size2index(tcache_max) + 1);
+		tsd_slow_update(tsd);
+		tsd_tcache_data_init(tsd);
+	} else if (old_tcache_max != tcache_max) {
+		/*
+		 * Still set thread_tcache_max and thread_nhbins even if
+		 * the tcache is not available yet because the values are
+		 * stored in tsd_t and are always available for changing.
+		 */
+		tsd_thread_tcache_max_set(tsd, tcache_max);
+		thread_nhbins_set(tsd, sz_size2index(tcache_max) + 1);
+		tsd_slow_update(tsd);
+	}
+	assert(tsd_thread_nhbins_get(tsd) == sz_size2index(tcache_max) + 1);
+}
+
 JEMALLOC_ALWAYS_INLINE bool
-tcache_small_bin_disabled(szind_t ind, cache_bin_t *bin) {
+tcache_small_bin_disabled(szind_t ind, cache_bin_t *bin,
+    unsigned current_nhbins) {
 	assert(ind < SC_NBINS);
-	bool ret = (cache_bin_info_ncached_max(&tcache_bin_info[ind]) == 0);
+	bool ret = (bin == NULL) ||
```

Comment:
For small bins, we allocate the bin struct regardless if they are enabled. So let's always pass in the bin here? There's only 1 place passing in NULL and I'm not sure if that's necessary.